### PR TITLE
Add csr helpers func

### PR DESCRIPTION
**Pull request by qiujian16**
Signed-off-by: Jian Qiu <jqiu@redhat.com>

### CODE REVIEW

**Comment by qiujian16** [comment]
/hold
still need to add unit test

**Comment by xuezhaojun** [comment]
/approve 
Looks good to me.

File: pkg/utils/csr_helpers.go [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+	"k8s.io/klog/v2"
+	"open-cluster-management.io/addon-framework/pkg/agent"
+	addonapiv1alpha1 "open-cluster-management.io/api/addon/v1alpha1"
+	clusterv1 "open-cluster-management.io/api/cluster/v1"
+)
+
+// DefaultSigner generates a signer func for addon agent to sign the csr using caKey and caData.
+func DefaultSigner(caKey, caData []byte) agent.CSRSignerFunc {
+	return func(csr *certificatesv1.CertificateSigningRequest) []byte {
+		blockTlsCrt, _ := pem.Decode(caData) // note: the second return value is not error for pem.Decode; it's ok to omit it.
```
Comment (yue9944882) [inline pkg/utils/csr_helpers.go:27]: > // note: the second return value is not error for pem.Decode; it's ok to omit it.
Write Preview

can you elaborate why the error is neglectable? what if the passing caData is not a valid base64 data?

File: pkg/utils/csr_helpers.go [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+		certs, err := x509.ParseCertificates(blockTlsCrt.Bytes)
+		if err != nil {
+			klog.Errorf("Failed to parse cert: %v", err)
+			return nil
+		}
+
+		blockTlsKey, _ := pem.Decode(caKey)
+		key, err := x509.ParsePKCS1PrivateKey(blockTlsKey.Bytes)
```
Comment (yue9944882) [inline pkg/utils/csr_helpers.go:38]: lets clarify in the comment that:

```
// For now only PKCS#1 is supported which assures the private key algorithm is RSA.
// TODO: Compatiblity w/ PKCS#8 key e.g. EC algorithm
```

File: pkg/utils/csr_helpers.go [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+	}
+}
+
+func signCSR(csr *certificatesv1.CertificateSigningRequest, caCert *x509.Certificate, caKey *rsa.PrivateKey) ([]byte, error) {
+	var usages []string
+	for _, usage := range csr.Spec.Usages {
+		usages = append(usages, string(usage))
+	}
+
+	certExpiryDuration := 365 * 24 * time.Hour
```
Comment (yue9944882) [inline pkg/utils/csr_helpers.go]: how about adding another public method `DefaultSignerWithExpiry` that allows custom expiry duration?

File: pkg/utils/csr_helpers.go (hunk 1 above)
Comment (xuezhaojun) [reply pkg/utils/csr_helpers.go:27]:
``` golang
// Decode will find the next PEM formatted block (certificate, private key
// etc) in the input. It returns that block and the remainder of the input. If
// no PEM data is found, p is nil and the whole of the input is returned in
// rest.
func Decode(data []byte) (p *Block, rest []byte) {
...
```
The second return value is `[]byte` type which we don't need in next steps. And If caData is not a valid base64 data, then the `blockTlsCrt` would be nil.

File: pkg/utils/csr_helpers.go [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+	}
+	policy := &config.Signing{
+		Default: &config.SigningProfile{
+			Usage:        usages,
+			Expiry:       certExpiryDuration,
+			ExpiryString: certExpiryDuration.String(),
+		},
+	}
+
+	cfs, err := local.NewSigner(caKey, caCert, signer.DefaultSigAlgo(caKey), policy)
```
Comment (yue9944882) [inline pkg/utils/csr_helpers.go]: personally, am not fan of involving external (cfssl) dependency to sign the certificate b/c they're already available in the go sdk. consider following the sample [here](https://github.com/open-cluster-management-io/cluster-proxy/blob/297e092ffcb13fbc7192b1d7b6c1dd2c8f4f15e7/pkg/addon/agent/agent.go#L128-L173)

File: pkg/utils/csr_helpers.go (hunk 1 above)
Comment (yue9944882) [reply pkg/utils/csr_helpers.go:27]: oh i thought its an error

File: pkg/utils/csr_helpers.go (hunk 3 above)
Comment (qiujian16) [reply pkg/utils/csr_helpers.go]: yes, just change to DefaultSignerWithExpiry.

**Review by yue9944882 — COMMENTED** [review]
/lgtm

**Pull request merged**